### PR TITLE
Improve performance of variable length patterns

### PR DESCRIPTION
**Pull request by systay**
Micro-optimisations of variable length expansion

### CODE REVIEW

**Comment by systay** [comment]
This replaces #7403

**Comment by boggle** [comment]
+1 on merging this

File: community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/pipes/VarLengthExpandPipe.scala [hunk 1]
```diff
@@ -52,29 +54,60 @@ case class VarLengthExpandPipe(source: Pipe,
...
+
+    new Iterator[(Node, Vector[Relationship])] {
+      override def next(): (Node, Vector[Relationship]) = {
+        val (node, rels: Vector[Relationship]) = stack.remove(stack.length - 1)
+        if (rels.length < maxDepthValue && nodeFilter(node)) {
+          val relationships: Iterator[Relationship] = state.query
+            .getRelationshipsForIds(node, dir, types.types(state.query))
+            .filter(filteringStep.filterRelationship(row, state))
+
+          val set: ContainsRelationship = if (rels.length > HASH_SET_CUTOFF) {
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/pipes/VarLengthExpandPipe.scala:96]: Is this really a net win? Doesn't this mean that in order to achieve O(1) `contains` we do a `clear` + `addAll` for each contains? I would have guessed that would be slower that only going through the `Vector` once checking for the existence of some element?

Comment (jexp) [inline community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/pipes/VarLengthExpandPipe.scala:96]: Yes, we measured it. @systay was sceptical first too. but then the numbers convinced him.

Comment (jexp) [inline community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/pipes/VarLengthExpandPipe.scala:96]: You should at some point have a look how the `Vector.contains()` works internally, it's really bad complexity :( and shows up as 90% cpu in profiling.

Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/pipes/VarLengthExpandPipe.scala:96]: Wouldn't it then be better to always use an `ArrayBuffer` (or even a `java.util.ArrayList` for that matter) ? I mean the only operation we use on the `Vector` is `contains`

Comment (jexp) [inline community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/pipes/VarLengthExpandPipe.scala:96]: We need an immutable structure here. As each added relationship should point to a new list which shouldn't affect the others, so we would have to copy both ArrayBuffer and ArrayList. 

And both ArrayBuffer and ArrayList have a linear contains complexity too.

Comment (jexp) [inline community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/pipes/VarLengthExpandPipe.scala:96]: @pontusmelke ?

**Pull request merged**